scipy/_lib/_array_api.py

```diff
@@ -0,0 +1,135 @@
+            raise TypeError("'numpy.matrix' are not supported")
+
+
+def array_namespace(*arrays, single_namespace=True):
+    compliance_scipy(*arrays)
+
+    if not _GLOBAL_CONFIG["SCIPY_ARRAY_API"]:
+        return np
+
+    # if we cannot get the namespace, np is used
+    # here until moved upstream
+    namespaces = set()
+    for array in arrays:
+        try:
+            namespaces.add(array_api_compat.array_namespace(array))
+        except TypeError:
+            namespaces.add(array_api_compat.numpy)
+
+    if single_namespace and len(namespaces) != 1:
+        raise ValueError(
+            f"Expected a single common namespace for array inputs, \
+              but got: {[n.__name__ for n in namespaces]}"
+        )
+
+    (xp,) = namespaces
+
+    return xp
+
+
+def asarray(array, dtype=None, order=None, copy=None, *, xp=None):
+    """Drop-in replacement for `np.asarray`.
+
+    Memory layout parameter `order` is not exposed in the Array API standard.
+    `order` is only enforced if the input array implementation
+    is NumPy based, otherwise `order` is just silently ignored.
+
+    The purpose of this helper is to make it possible to share code for data
+    container validation without memory copies for both downstream use cases
+    """
+    if xp is None:
+        xp = array_namespace(array)
+    if xp.__name__ in {"numpy", "array_api_compat.numpy", "numpy.array_api"}:
+        # Use NumPy API to support order
+        if copy is True:
+            array = np.array(array, order=order, dtype=dtype)
+        else:
+            array = np.asarray(array, order=order, dtype=dtype)
+
+        # At this point array is a NumPy ndarray. We convert it to an array
+        # container that is consistent with the input's namespace.
+        return xp.asarray(array)
+    else:
+        return xp.asarray(array, dtype=dtype, copy=copy)
+
+
+def asarray_namespace(*arrays):
+    """Validate and convert arrays to a common namespace.
+
+    Parameters
+    ----------
+    *arrays : sequence of array_like
+        Arrays to validate and convert.
+
+    Returns
+    -------
+    *arrays : sequence of array_like
+        Validated and converted arrays to the common namespace.
+    namespace : module
+        Common namespace.
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> x, y, xp = asarray_namespace([0, 1, 2], np.arange(3))
+    >>> xp.__name__
+    'array_api_compat.numpy'
+    >>> x, y
+    (array([0, 1, 2]]), array([0, 1, 2]))
+
+    """
+    arrays = list(arrays)  # probably not good
+    xp = array_namespace(*arrays)
+
+    for i, array in enumerate(arrays):
+        arrays[i] = asarray(array, xp=xp)
+
+    return *arrays, xp
+
+
+def to_numpy(array, xp):
+    """Convert `array` into a NumPy ndarray on the CPU.
+
+    This is specially useful to pass arrays to Cython.
+    """
+    xp_name = xp.__name__
+
+    if xp_name in {"array_api_compat.torch", "torch"}:
+        return array.cpu().numpy()
```

Comment:
You don't want to do this. Just using `np.asarray(array)` is better. As a rule, never do silent device transfers from GPU to CPU or vice verse. This is true for CuPy too - that should simply error.

> This is specially useful to pass arrays to Cython.

This conversation was specifically about PyTorch CPU tensors. There may be other cpu array libraries where this works. But basically, this function doesn't seem needed, in any code where you planned to use this I think you want `np.asarray` instead.

Reply:
Ok I will remove 👍 This was in sklearn so I thought this was more "optimal" to do the conversion using that.

Reply:
Hmm interesting. @thomasjpfan I thought you preferred exceptions?

Reply:
Scikit-learn has a function to do device transfer strictly for testing purposes and is not part of public API. Scikit-learn checks that the GPU implementation gives similar results as the CPU implementation.

This type of testing does not have full coverage, but it's nice to have some checks to make sure the GPU arrays does something reasonable.

Reply:
Ah yes, for testing it makes sense indeed. As long as we figure out how to make sure to not accidentally introduce it into the code base outside of testing.

Reply:
I added some utilities for testing back on the host from the device in my branch.. my approach was quite different though.. perhaps I can try to make a PR to this branch but...